// @histoire/controls/.histoire/dist/assets/HstNumber.story.1e1964ff.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
import { ax as _export_sfc, d as defineComponent, c as computed, r as ref, o as onUnmounted, a as openBlock, j as createBlock, k as withCtx, f as renderSlot, M as withDirectives, Z as vModelText, i as unref, e as createBaseVNode, x as mergeProps, P as isRef, n as normalizeClass, h as normalizeStyle, q as resolveComponent, l as createVNode } from "./vendor.2833ae3d.js";
import { H as HstWrapper } from "./HstWrapper.246b6982.js";
const __default__ = {
  name: "HstNumber",
  inheritAttrs: false
};
const _sfc_main$1 = /* @__PURE__ */ defineComponent(__spreadProps(__spreadValues({}, __default__), {
  props: {
    title: { type: String, required: false },
    modelValue: { type: Number, required: true }
  },
  emits: {
    "update:modelValue": (newValue) => true
  },
  setup(__props, { emit }) {
    const props = __props;
    const numberModel = computed({
      get: () => props.modelValue,
      set: (value) => {
        emit("update:modelValue", value);
      }
    });
    const input = ref();
    function focusAndSelect() {
      input.value.focus();
      input.value.select();
    }
    const isDragging = ref(false);
    let startX;
    let startValue;
    function onMouseDown(event) {
      isDragging.value = true;
      startX = event.clientX;
      startValue = numberModel.value;
      window.addEventListener("mousemove", onMouseMove);
      window.addEventListener("mouseup", stopDragging);
    }
    function onMouseMove(event) {
      let step = parseFloat(input.value.step);
      if (isNaN(step)) {
        step = 1;
      }
      numberModel.value = startValue + Math.round((event.clientX - startX) / 10 / step) * step;
    }
    function stopDragging() {
      isDragging.value = false;
      window.removeEventListener("mousemove", onMouseMove);
      window.removeEventListener("mouseup", stopDragging);
    }
    onUnmounted(() => {
      stopDragging();
    });
    return (_ctx, _cache) => {
      return openBlock(), createBlock(HstWrapper, {
        class: normalizeClass(["htw-cursor-ew-resize htw-items-center", [
          _ctx.$attrs.class,
          { "htw-select-none": isDragging.value }
        ]]),
        title: __props.title,
        style: normalizeStyle(_ctx.$attrs.style),
        onClick: focusAndSelect,
        onMousedown: onMouseDown
      }, {
        actions: withCtx(() => [
          renderSlot(_ctx.$slots, "actions")
        ]),
        default: withCtx(() => [
          withDirectives(createBaseVNode("input", mergeProps({
            ref_key: "input",
            ref: input
          }, __spreadProps(__spreadValues({}, _ctx.$attrs), { class: null, style: null }), {
            "onUpdate:modelValue": _cache[0] || (_cache[0] = ($event) => isRef(numberModel) ? numberModel.value = $event : null),
            type: "number",
            class: [{
              "htw-select-none": isDragging.value
            }, "htw-text-inherit htw-bg-transparent htw-w-full htw-outline-none htw-pl-2 htw-py-1 -htw-my-1 htw-border htw-border-solid htw-border-black/25 dark:htw-border-white/25 focus:htw-border-primary-500 dark:focus:htw-border-primary-500 htw-rounded-sm htw-cursor-ew-resize"]
          }), null, 16), [
            [
              vModelText,
              unref(numberModel),
              void 0,
              { number: true }
            ]
          ])
        ]),
        _: 3
      }, 8, ["title", "class", "style"]);
    };
  }
}));
_sfc_main$1.__file = "src/components/number/HstNumber.vue";
var HstNumber = /* @__PURE__ */ _export_sfc(_sfc_main$1, [["__file", "/Users/akryum/Projects/histoire/packages/histoire-controls/src/components/number/HstNumber.vue"]]);
const _sfc_main = /* @__PURE__ */ defineComponent({
  name: "HstNumber.story",
  setup(__props) {
    function initState() {
      return {
        count: 20,
        step: 1
      };
    }
    return (_ctx, _cache) => {
      const _component_Variant = resolveComponent("Variant");
      const _component_Story = resolveComponent("Story");
      return openBlock(), createBlock(_component_Story, { title: "HstNumber" }, {
        default: withCtx(() => [
          createVNode(_component_Variant, {
            title: "default",
            "init-state": initState
          }, {
            default: withCtx(({ state }) => [
              createVNode(HstNumber, {
                modelValue: state.count,
                "onUpdate:modelValue": ($event) => state.count = $event,
                step: state.step,
                title: "Count"
              }, null, 8, ["modelValue", "onUpdate:modelValue", "step"])
            ]),
            controls: withCtx(({ state }) => [
              createVNode(HstNumber, {
                modelValue: state.count,
                "onUpdate:modelValue": ($event) => state.count = $event,
                title: "Count"
              }, null, 8, ["modelValue", "onUpdate:modelValue"]),
              createVNode(HstNumber, {
                modelValue: state.step,
                "onUpdate:modelValue": ($event) => state.step = $event,
                title: "Step"
              }, null, 8, ["modelValue", "onUpdate:modelValue"])
            ]),
            _: 1
          })
        ]),
        _: 1
      });
    };
  }
});
_sfc_main.__file = "src/components/number/HstNumber.story.vue";
var HstNumber_story = /* @__PURE__ */ _export_sfc(_sfc_main, [["__file", "/Users/akryum/Projects/histoire/packages/histoire-controls/src/components/number/HstNumber.story.vue"]]);
export { HstNumber_story as default };
